Collect and send custom attributes for capture feature

satsukies opened the pull request:
I added an implementation that will be used to enhance the capture function provided after the next version.

- Users can add attributes freely and their values via `CustomAttributes` class .
- Attributes are divided into two groups. `BuildEnvironment` or `RuntimeExtra`.
  - DeployGate instance keeps `CustomAttriutes` of each groups.
- The attributes are sent automatically to DeployGate by SDK when the capture is created.
  - When SDK received `ACTION_COLLECT_DEVICE_STATES` event, SDK will write values via content provider.

satsukies commented on sdk/src/main/java/com/deploygate/sdk/DeployGate.java:
Please check plan of implementation before completing work. (interface design/naming/etc...)

satsukies commented on sdk/src/main/java/com/deploygate/sdk/DeployGate.java:
```diff
@@ -1390,4 +1401,192 @@ public static String getDistributionUserName() {
         return sInstance.mDistributionUserName;
     }
+
+    public static boolean putBuildEnvironmentValue(String key, String value) {
```
We need to add comments each public method.

satsukies commented:
@jmatsu 
Could you check if my implementation plan has any problems or concerns before finish my work?
Thank you.

jmatsu commented on sdk/src/main/java/com/deploygate/sdk/DeployGate.java:
```diff
@@ -59,6 +67,8 @@ public class DeployGate {
     private final CustomLogInstructionSerializer mCustomLogInstructionSerializer;
     private final HashSet<DeployGateCallback> mCallbacks;
     private final HashMap<String, Bundle> mPendingEvents;
+    private final HashMap<String, Object> mBuildEnvironments;
+    private final HashMap<String, Object> mRuntimeExtras;
```
Please use ConcurrentHashMap or LRUCache for the thread-safety.

jmatsu commented on sdk/src/main/java/com/deploygate/sdk/DeployGate.java:
```diff
@@ -1390,4 +1402,192 @@ public static String getDistributionUserName() {
+            return false;
+        }
+
+        if (value instanceof String && ((String) value).length() > MAX_USER_INPUT_VALUE_LENGTH) {
```
Please check `Object` types. String, Number, Boolean are only allowed, right?

satsukies replied:
> String, Number, Boolean are only allowed, right?

Yes.
I forgot to check the value is instance of Number or Boolean after this line.

jmatsu commented on sdk/src/main/java/com/deploygate/sdk/DeployGate.java:
```diff
@@ -1390,4 +1402,192 @@ public static String getDistributionUserName() {
 
         return sInstance.mDistributionUserName;
     }
+
```
I prefer providing `getBuildEnvironment(): SomeExtras` and `getRuntimeExtras(): SomeExtras` much more than providing many methods from DeployGate class.

`SomeExtras` above is kinda thread-safe KVS and has named interfaces like `putBoolean`, `putInt` and `putString`.

satsukies replied:
I understand. 
If we are provide with `Some Extra` class, we can also provide builder pattern for create values. 👍

jmatsu reviewed (commented):
The approach basically looks good. Please check my comments.

satsukies commented:
@jmatsu 
Thank you for your feedback.
I made some changes after your feedback.

1. Added `CustomAttributes` class for keeping custom attributes.
2. Added to checking key and value following the feature specs.
3. Added tests.
4. Added implementation of writing custom attributes to content provider using received URI.

Please re-review this PR. Thank you.

jmatsu commented on sdk/src/main/java/com/deploygate/sdk/CustomAttributes.java:
```diff
@@ -0,0 +1,115 @@
+      return false;
+    }
+
+    if (key.length() < MIN_KEY_LENGTH || key.length() > MAX_KEY_LENGTH || !VALID_KEY_PATTERN.matcher(key).matches()) {
```
Length validation is included by the regexp.

```suggestion
    if (!VALID_KEY_PATTERN.matcher(key).matches()) {
```

jmatsu reviewed (changes requested):
The approach looks basically good. Please check my comments.

jmatsu commented:
@satsukies 
I'm sorry for missing another point. Please use https://github.com/DeployGate/deploygate-android-sdk/blob/250ed74e93320d10ae710633c34fc8e85007270a/sdk/src/main/java/com/deploygate/sdk/internal/Logger.java for new logging!

satsukies commented on sdkMock/src/test/java/com/deploygate/sdk/CustomAttributesInterfaceTest.java:
use the same test located in sdk module to checking provide same methods each modules.

satsukies commented:
@jmatsu 
Thanks for your careful feedback!
Please re-review some fixes and whole codes.

jmatsu reviewed (changes requested):
Please be careful for synchronization, `sdk-mock` interfaces and AIDL operations.

satsukies commented on sdk/src/test/java/com/deploygate/sdk/CustomAttributesInterfaceTest.java:
```diff
@@ -0,0 +1,68 @@
+  }
+
+  @Test
+  public void putString() {
```
Part of these testcases you said are maybe covered at CustomAttributesTest.java 👀  

> - Cannot put values if it exceeds the max size.
> - Can put values again if the size has been decreased.

Sorry if you already seen it, please take a look below. 
https://github.com/DeployGate/deploygate-android-sdk/blob/c8296588cddc3eca4ee56ce779c631e4764aa8e6/sdk/src/test/java/com/deploygate/sdk/CustomAttributesTest.java

> Check if it's empty with filled attributes.

I see, add testcase of checking it 👍

jmatsu replied:
Oh sorry, you are right. They are tested in another test file.

satsukies commented on sdkMock/src/main/java/com/deploygate/sdk/DeployGate.java:
```diff
@@ -179,4 +179,12 @@ public static void composeComment(String defaultComment) {
     }
+
+    public static CustomAttributes getBuildEnvironment() {
+        return null;
```
This comment remind me of you said "must be always non-null for friendly interfaces" in previous review.
I'll fix it.

satsukies commented on sdk/src/main/java/com/deploygate/sdk/DeployGate.java:
```diff
@@ -53,8 +56,9 @@ public class DeployGate {
-    private static CustomAttributes sRuntimeExtra = null;
+    private static CustomAttributes sBuildEnvironment;
+    private static CustomAttributes sRuntimeExtra;
+    private static CustomAttributes sSdkRuntimeExtra;
```
This attrs will be used to collect data for valuable on the debugging.
It is always empty at this point.

satsukies commented on sdk/src/main/java/com/deploygate/sdk/DeployGate.java:
```diff
@@ -139,16 +143,43 @@ public void onEvent(
+                synchronized (getRuntimeExtra().attributes) {
+                    if (!getRuntimeExtra().attributes.isEmpty()) {
+                        for (Map.Entry<String, Object> attr : getRuntimeExtra().attributes.entrySet()) {
+                            String userKey = String.format("%s.%s", mHostApp.packageName, attr.getKey());
```
We need to append prefix to attribute keys created by user.

jmatsu replied:
Please don't add the extra time-consuming processing. If we append the prefix in CustomAttributes, doesn't it work?

satsukies commented on sdk/src/test/java/com/deploygate/sdk/CustomAttributesTest.java:
```diff
@@ -97,24 +96,49 @@ public void put_remove_removeAll_maxSize() {
-    } catch (Exception e) {
-      Truth.assertWithMessage("Failed to parse JSON").fail();
-    }
+  public void size() {
```
add new testcase for checking behavior size() and isEmpty() methods.

satsukies commented:
@jmatsu 
Thanks for your review!
So sorry to request reviews many times,  could you please re-review?

jmatsu commented on sdkMock/src/test/java/com/deploygate/sdk/CustomAttributesInterfaceTest.java:
```diff
@@ -0,0 +1 @@
+../../../../../../../sdk/src/test/java/com/deploygate/sdk/CustomAttributesInterfaceTest.java
```
👍

jmatsu commented on sdkMock/src/main/java/com/deploygate/sdk/DeployGate.java:
```diff
@@ -5,6 +5,8 @@
 
 public class DeployGate {
 
+    private static final CustomAttributes sAttributes = new CustomAttributes();
```
👍

jmatsu commented on sdk/src/test/java/com/deploygate/sdk/DeployGateInterfaceTest.java:
```diff
@@ -364,4 +364,14 @@ public void composeComment__String() {
+
+    @Test
+    public void getBuildEnvironment() {
+        DeployGate.getBuildEnvironment();
```
Please add *assertions* to testcases of functions that return values. Calling this function does not definitely mean that they never return null.

jmatsu commented on sdk/src/test/java/com/deploygate/sdk/CustomAttributesTest.java:
```diff
@@ -0,0 +1,144 @@
+    attributes.putString("key6", "value");
+    attributes.putString("key7", "value");
+    attributes.putString("key8", "value");
+    Truth.assertThat(attributes.size()).isEqualTo(8);
```
You missed a testcase for checking how *overwriting* existing values work.

jmatsu commented on sdk/src/test/java/com/deploygate/sdk/CustomAttributesTest.java:
```diff
@@ -0,0 +1,144 @@
+  }
+
+  @Test
+  public void put__accept_when_valid_key() {
```
Please use different values for each key for the proper testing.

jmatsu commented on sdk/src/test/java/com/deploygate/sdk/CustomAttributesTest.java:
```diff
@@ -0,0 +1,144 @@
+  }
+
+  @Test
+  public void not_exceed_max_size() {
```
ditto: use different values.

jmatsu commented on sdk/src/test/java/com/deploygate/sdk/CustomAttributesInterfaceTest.java:
```diff
@@ -0,0 +1,68 @@
+
+  @Test
+  public void putString() {
+    attributes.putString("key", "value");
```
The signature is consist of *the name*, *the arguments*, *the modifiers* and **the return type**. 

It may be difficult to check the modifiers, but it's easy to check if the returned value is Boolean. Could you please add assertions?

jmatsu commented on sdk/src/main/java/com/deploygate/sdk/DeployGate.java:
```diff
@@ -127,6 +138,48 @@ public void onEvent(
+
+                ContentValues cv = new ContentValues();
+                synchronized (getBuildEnvironment().attributes) {
+                    if (!getBuildEnvironment().attributes.isEmpty()) {
+                        String buildEnvironmentJson = new JSONObject(getBuildEnvironment().attributes).toString();
+                        cv.put(DeployGateEvent.ATTRIBUTE_KEY_BUILD_ENVIRONMENT, buildEnvironmentJson);
+                    }
```
What makes you check if it's empty? Can you explain what will happen in this synchronized block?

It looks a meaningless synchronized block because an empty CustomAttributes always produces the same result, i.e. an empty json object, with the optimized synchronization via *ConcurrentHashMap*. So, you should just check if the json string is `{}` or not.

jmatsu reviewed (changes requested):
Please be careful about the device resources.

satsukies commented:
We discussed in-person to plan for how to fix current codes.
Following the discussion, I worked on belows.

- Separating params of sending attributes from SDK by 3 groups. BuildEnvironment, RuntimeExtras and SdkDeviceStatus.
- Improve implementation of  CustomAttributes to be thread-safe.

@jmatsu 
Could you please check my codes? Thanks.

jmatsu reviewed (changes requested):
Almost LGTM. Could you please add one important testcase?

satsukies commented on sdk/src/test/java/com/deploygate/sdk/CustomAttributesTest.java:
```diff
@@ -0,0 +1,163 @@
+import java.util.concurrent.Executors;
+
+@RunWith(AndroidJUnit4.class)
+public class CustomAttributesTest {
```
Oh sorry.
I added testcase for `getJSONString` method in 063fc76 .

jmatsu reviewed (approved):
Great. Thank you for the tough work.

satsukies commented:
Thanks for comment. 
I added important testcase for `getJSONString` method.

satsukies commented:
> Thank you for the tough work.

You too, thank you for your review many times and careful feedback.
